3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/MailgunAdapter.scala

```diff
@@ -0,0 +1,205 @@
+ */
+object MailgunAdapter extends Adapter {
+
+  // Vendor name for Failure Message
+  private val VendorName = "Mailgun"
+
+  // Tracker version for an Mailgun Tracking webhook
+  private val TrackerVersion = "com.mailgun-v1"
+
+  // Expected content type for a request body
+  private val ContentTypes = List("application/x-www-form-urlencoded", "multipart/form-data")
+  private val ContentTypesStr = ContentTypes.mkString(" or ")
+
+  // Schemas for reverse-engineering a Snowplow unstructured event
+  private val EventSchemaMap = Map (
+    "bounced"      -> SchemaKey("com.mailgun", "message_bounced", "jsonschema", "1-0-0").toSchemaUri,
+    "clicked"      -> SchemaKey("com.mailgun", "message_clicked", "jsonschema", "1-0-0").toSchemaUri,
+    "complained"   -> SchemaKey("com.mailgun", "message_complained", "jsonschema", "1-0-0").toSchemaUri,
+    "delivered"    -> SchemaKey("com.mailgun", "message_delivered", "jsonschema", "1-0-0").toSchemaUri,
+    "dropped"      -> SchemaKey("com.mailgun", "message_dropped", "jsonschema", "1-0-0").toSchemaUri,
+    "opened"       -> SchemaKey("com.mailgun", "message_opened", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribed" -> SchemaKey("com.mailgun", "recipient_unsubscribed", "jsonschema", "1-0-0").toSchemaUri
+  )
+
+  /**
+   * Converts a CollectorPayload instance into raw events.
+   *
+   * A Mailgun Tracking payload contains one single event
+   * in the body of the payload, stored within a HTTP encoded
+   * string.
+   *
+   * @param payload The CollectorPayload containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation. Not used
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents = {
+    (payload.body, payload.contentType) match {
+      case (None, _)                                                => s"Request body is empty: no ${VendorName} events to process".failNel
+      case (_, None)                                                => s"Request body provided but content type empty, expected ${ContentTypesStr} for ${VendorName}".failNel
+      case (_, Some(ct)) if !ContentTypes.exists(ct.startsWith(_))  => s"Content type of ${ct} provided, expected ${ContentTypesStr} for ${VendorName}".failNel
+      case (Some(body), _) if (body.isEmpty)                        => s"${VendorName} event body is empty: nothing to process".failNel
+      case (Some(body), Some(ct))                                   => {
+        val params = toMap(payload.querystring)
+        try {
+          val bodyMap = getBoundary(ct)
+                          .map(parseMultipartForm(body, _))
+                          .getOrElse(toMap(URLEncodedUtils.parse(URI.create("http://localhost/?" + body), "UTF-8").toList))
```

Comment:
if we want to be consistent with other parts of the codebase this will look like:

```scala
val bodyMap = getBoundary(ct)
  .map(parseMultipartForm(body, _))
  .getOrElse(toMap(URLEncodedUtils.parse(URI.create("http://localhost/?" + body), "UTF-8").toList))
```